### PR TITLE
Support ecma

### PR DESCRIPTION
**Pull request by AtofStryker**
Essentially cut over the the webpack loader to use ES6 Modules with working tests. Loader (compiler) is now transpiled through babel to support commonjs. Added a circle ci job to test to kick off on code pushes and added linting. README is updated to accurately reflect usage and requirements.

Eventually I will make a PR to get this as a feature in the original library via a flag.

I am also starting with a `beta` canary on publish (once I have access) to test functionality outside `npm link`

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -1,9 +0,0 @@
-root = true
```
Comment (racoleman) [inline .editorconfig:1]: Why remove the `.editorconfig`?

File: .eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (racoleman) [inline .eslintrc.json:1]: Can we keep this config the same as grove core? Having differing styles between repos can be frustrating.

File: README.md [hunk 3]
```diff
@@ -41,7 +42,7 @@ Config example using a querystring:
 module.exports = {
   module: {
     loaders: [
-      { test: /\.ejs$/, loader: 'ejs-loader?variable=data' },
+      { test: /\.ejs$/, loader: '@groveco/ejs-loader?variable=data' },
```
Comment (racoleman) [inline README.md:45]: This entire section should be removed, as passing options via querystring was deprecated in Webpack 3.

File: README.md [hunk 4]
```diff
@@ -57,7 +58,7 @@ module.exports = {
         loaders: [
             {
                 test: /\.ejs$/,
-                loader: 'ejs-loader',
+                loader: '@groveco/ejs-loader',
```
Comment (racoleman) [inline README.md:61]: The `loaders` configuration is deprecated in favor of `rules`.

File: README.md [hunk 5]
```diff
@@ -110,16 +111,19 @@ Lodash template function does not provide `include` method of [ejs module](http:
 
 As a result, `renderedHtml` becomes a string `<h1><a href="http://example.com">Example</a></h1>`.
 
-
-
 ## Release History
-* 1.0.0 - Changed export format to ESM
+* 1.0.0-beta.1 - Inital fork of [ejs-loader](https://github.com/difelice/ejs-loader) with support for ECMA Modules and deprecating CommonsJS
```
Comment (racoleman) [inline README.md:115]: Unless you're anticipating issues, I would skip the beta tag. Otherwise, we lose out on automatic patches and will have to have corresponding update PRs in grove core for minor bug fixes.

File: README.md [hunk 6]
```diff
@@ -1,27 +1,28 @@
...
 
 ``` javascript
-var template = require("ejs!./file.ejs");
+import template from '@groveco/ejs-loader?variable=data!./file.ejs';
 // => returns the template function compiled with undesrcore (lodash) templating engine.
 
 // And then use it somewhere in your code
 template(data) // Pass object with data
 ```
+The variable option is `required` to compile EJS templates into ES compatible modules. If the `variable` option is not provided as a loader or `query` [option](https://webpack.js.org/concepts/loaders/#loader-features), an `UnsupportedConfiguration` Exception will be thrown throw. Please see https://github.com/lodash/lodash/issues/3709#issuecomment-375898111 for additional details
```
Comment (racoleman) [inline README.md:25]: If the option is required, why don't we give it a default instead of throwing an exception?

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            - "node_modules"
+  build:
+    docker:
+      - image: circleci/node:lts
+    steps:
+      - checkout
+      - restore_cache:
+          keys:
+            - v1-npm-dependencies-{{ checksum "package-lock.json" }}
+      - run: npm run build
```
Comment (racoleman) [inline .circleci/config.yml:26]: Why do we need to build anything? If the loader is written to a level that Node and Jest support, you won't need Babel or a build step at all.

**Review by racoleman — COMMENTED** [review]
This looks pretty comprehensive; definitely in favor of the CircleCI integration. That said, I would split out any work you consider critical to HHR-129 into a separate PR, so we can focus on that. It's a blocker to the code splitting work on grove core.

File: index.js [hunk 8]
```diff
@@ -1,25 +1,35 @@
...
+import _ from 'lodash'
+import loaderUtils from 'loader-utils'
 
-function getOptions(context) {
+class UnsupportedConfiguration extends Error {}
+
+function getOptions (context) {
   if (context.options && context.options.ejsLoader) {
-    return context.options.ejsLoader;
+    return context.options.ejsLoader
```
Comment (jessewagstaff) [inline index.js:8]: Why the change to drop semicolons?

File: README.md (hunk 6 above)
Comment (AtofStryker) [reply README.md:25]: @racoleman only reason I am opting for the exception here is it might not be clear how the default value plays into a given template. I want it to be clear that a value must be explicitly passed in and generic templates will not be supported, as opposed to the template being compiled without errors but at runtime syntax errors are potentially thrown.

File: .circleci/config.yml (hunk 7 above)
Comment (AtofStryker) [reply .circleci/config.yml:26]: @racoleman Two problems: 

If `index.js` is written as a ES Module, webpack uses [interpret](https://www.npmjs.com/package/interpret) under the hood, which uses `cjs-loader`. If you have babel set up locally, you can get away with `webpack.config.babel.js` but not every user will be digesting this with `babel`. 

Jest (because its mocking everyone 😆 ) uses its own [resolver](https://github.com/facebook/jest/blob/master/packages/jest-runtime/src/index.ts#L56) which doesn't support ES6 import statements.

I figured since we are using es6 all the way here, that it would make sense to use babel for the tests and use babel to transpile the dist directory to avoid these issue.

File: index.js (hunk 8 above)
Comment (AtofStryker) [reply index.js:8]: @jessewagstaff I dropped them from the linter because we are now using babel to transform the file. They aren't required by default and are added in the bundle.

File: README.md (hunk 5 above)
Comment (AtofStryker) [reply README.md:115]: @racoleman this is just to test the package out. Will bump to 1.0.0 so we can get patch support once ready

File: README.md (hunk 3 above)
Comment (AtofStryker) [reply README.md:45]: @racoleman it's deprecated but technically someone could use this package with older versions of webpack. My guess is that if people are still using `ejs` templates, they could also be using older versions of webpack

File: README.md (hunk 4 above)
Comment (AtofStryker) [reply README.md:61]: @racoleman almost a similar deal [here](https://github.com/groveco/ejs-loader/pull/1#discussion_r431982953), but at a minimum we should probably add documentation for the rules configuration

File: .editorconfig (hunk 1 above)
Comment (AtofStryker) [reply .editorconfig:1]: @racoleman I think I did it by mistake 😳 . Will readd

File: .eslintrc.json (hunk 2 above)
Comment (AtofStryker) [reply .eslintrc.json:1]: @racoleman we can but we do not need nearly as many rules as we do in grove core. I opted for a simpler config for that reason, as well as rules that are more fitting for a node environment IMO

File: .circleci/config.yml (hunk 7 above)
Comment (racoleman) [reply .circleci/config.yml:26]: That's because [Webpack loaders should not be written as ES modules ](https://webpack.js.org/contribute/writing-a-loader/)😉.

Now you might could work around using a build for Jest by using `babel-jest`, but I have to ask: is it worth it to add all this infrastructure just to use import statements in one file?

File: .eslintrc.json (hunk 2 above)
Comment (racoleman) [reply .eslintrc.json:1]: Simpler I understand, but we should keep things consistent (e.g. semicolons, as @jessewagstaff pointed out).

File: README.md (hunk 3 above)
Comment (racoleman) [reply README.md:45]: If you wish to maintain documentation for older versions, I'd suggest putting this under a "Webpack 2" header and make sure the first example is written to Webpack 4.

File: .eslintrc.json (hunk 2 above)
Comment (AtofStryker) [reply .eslintrc.json:1]: @racoleman I don't know if I agree. The target here is different and some of the options here are preferred in the node.js environment. I think we should opt for it. Consistency for the sake of consistency doesn't matter, especially when it adds complexity and bringing over potential baggage from other or older styles and practices. It's the end goal that matters. 

This is also why I don't think it's in this package's best interest to be maintained by the grove organization. This is a general open source project were people can contribute outside of grove. Simpler in this case is almost always better.

File: .circleci/config.yml (hunk 7 above)
Comment (AtofStryker) [reply .circleci/config.yml:26]: > That's because [Webpack loaders should not be written as ES modules ](https://webpack.js.org/contribute/writing-a-loader/)😉.

I don't think that is implied anywhere. It actually implies the opposite quite a bit in the documentation, such as in [loader-utils][https://webpack.js.org/contribute/writing-a-loader/#loader-utilities], [loader-dependencies](https://webpack.js.org/contribute/writing-a-loader/#loader-utilities)

They also recommend using `babel-jest` for testing and having your module src be in ES6, but having a runtime that uses CommonJS. I figured the best way to accomplish that was to transpile the build 

> Now you might could work around using a build for Jest by using babel-jest, but I have to ask: is it worth it to add all this infrastructure just to use import statements in one file?

yes, ES6 is life! 😆  Honestly don't think it adds much but I understand the rationale

**Comment by AtofStryker** [comment]
@jessewagstaff spoke with @racoleman about these changes. Since I no longer need the es6 imports for the test, I am going to remove babel and the transpilation since we really don't need it. I will also remove the linter and update the documentation to be webpack 4+
